modules/tensor_mechanics/src/kernels/StressDivergence2DTensors.C

```diff
@@ -0,0 +1,284 @@
+        _Jacobian_mult[_qp], 2, jvar, test_z, phi); // test_z and phi_y
+
+    first_term = first_sum + mixed_sum2;
+  }
+  else if (ivar == 1 && jvar == 0)
+  {
+    const Real second_sum = ElasticityTensorTools::elasticJacobian(
+        _Jacobian_mult[_qp], ivar, jvar, test, phi); // test_y and phi_x
+    const Real mixed_sum1 = ElasticityTensorTools::elasticJacobian(
+        _Jacobian_mult[_qp], ivar, 2, test, phi_z); // test_y and phi_z
+
+    first_term = second_sum + mixed_sum1;
+  }
+  else if (ivar == 1 && jvar == 1)
+    first_term = ElasticityTensorTools::elasticJacobian(
+        _Jacobian_mult[_qp], ivar, jvar, test, phi); // test_y and phi_y
+  else
+    mooseError("Invalid component in Jacobian Calculation");
+
+  Real val = 0.0;
+  // volumetric locking correction
+  // K = Bbar^T_i * C * Bbar^T_j where Bbar = B + Bvol
+  // K = B^T_i * C * B_j + Bvol^T_i * C * Bvol_j + B^T_i * C * Bvol_j + Bvol^T_i * C * B_j
+  if (_volumetric_locking_correction)
+  {
+    RealGradient new_test(2, 0.0);
+    RealGradient new_phi(2, 0.0);
+
+    new_test(0) = _grad_test[_i][_qp](0);
+    new_test(1) = _grad_test[_i][_qp](1);
+    new_phi(0) = _grad_phi[_j][_qp](0);
+    new_phi(1) = _grad_phi[_j][_qp](1);
+
+    // Bvol^T_i * C * Bvol_j
+    Real sum = 0.0;
+    for (unsigned i = 0; i < 2; ++i)
+      for (unsigned j = 0; j < 2; ++j)
+        sum += _Jacobian_mult[_qp](i, i, j, j);
+    val += sum * (_avg_grad_test[_i][ivar] - new_test(ivar)) *
+           (_avg_grad_phi[_j][jvar] - new_phi(jvar)) / 2.0;
+
+    // B^T_i * C * Bvol_j
+    RealGradient sum_2x1;
+    sum_2x1(0) = _Jacobian_mult[_qp](0, 0, 0, 0) + _Jacobian_mult[_qp](0, 0, 1, 1);
+    sum_2x1(1) = _Jacobian_mult[_qp](1, 1, 0, 0) + _Jacobian_mult[_qp](1, 1, 1, 1);
+    if (ivar == 0 && jvar == 0)
+      val += sum_2x1(0) * test(0) * (_avg_grad_phi[_j][0] - new_phi(0));
+    else if (ivar == 0 && jvar == 1)
+      val += sum_2x1(0) * test(0) * (_avg_grad_phi[_j][1] - new_phi(1));
+    else if (ivar == 1 && jvar == 0)
+      val += sum_2x1(1) * test(1) * (_avg_grad_phi[_j][0] - new_phi(0));
+    else
+      val += sum_2x1(1) * test(1) * (_avg_grad_phi[_j][1] - new_phi(1));
+
+    // Bvol^T_i * C * B_j
+    // val = trace (C * B_j) *(avg_grad_test[_i][ivar] - new_test(ivar))
+    if (jvar == 0)
+      for (unsigned int i = 0; i < 2; ++i)
+        val +=
+            (_Jacobian_mult[_qp](i, i, 0, 0) * phi(0) + _Jacobian_mult[_qp](i, i, 0, 1) * phi(1)) *
+            (_avg_grad_test[_i][ivar] - new_test(ivar));
+    else if (jvar == 1)
+      for (unsigned int i = 0; i < 2; ++i)
+        val +=
+            (_Jacobian_mult[_qp](i, i, 0, 1) * phi(0) + _Jacobian_mult[_qp](i, i, 1, 1) * phi(1)) *
+            (_avg_grad_test[_i][ivar] - new_test(ivar));
+  }
+
+  return val / 2.0 + first_term;
+}
+
+void
+StressDivergence2DTensors::computeAverageGradientTest()
+{
+  // calculate volume averaged value of shape function derivative
+  _avg_grad_test.resize(_test.size());
```

Comment:
I don't believe it can be done much earlier.  _test and _phi have size zero at construction time.  I'll move these to precalculate methods, which will help.